File: python/examples/hello_sub.py
#!/usr/bin/env python3
"""Subscribe to HelloWorld samples on domain 0.

Usage:
    python examples/hello_sub.py [--domain DOMAIN_ID]

Receives HelloWorld messages from any publisher on the domain.
"""

import sys
import os
import argparse
import logging

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from vibedds.participant import DomainParticipant
from vibedds.qos import QosPolicy, ReliabilityKind
from vibedds.type_support import HelloWorldType


def on_data(payload: bytes):
    try:
        message = HelloWorldType.deserialize(payload)
        print(f"Received: {message}")
    except Exception as e:
        print(f"Failed to deserialize: {e}")


def main():
    parser = argparse.ArgumentParser(description="HelloWorld Subscriber")
    parser.add_argument("--domain", type=int, default=0)
    parser.add_argument("-v", "--verbose", action="store_true")
    args = parser.parse_args()

    logging.basicConfig(
        level=logging.DEBUG if args.verbose else logging.INFO,
        format="%(asctime)s %(name)s %(levelname)s: %(message)s",
    )

    dp = DomainParticipant(domain_id=args.domain)
    dp.start()

    topic = dp.create_topic("HelloWorldTopic", HelloWorldType.TYPE_NAME)
    reader = dp.create_reader(topic, QosPolicy(reliability=ReliabilityKind.BEST_EFFORT))
    reader.on_data(on_data)

    print(f"HelloWorld Subscriber started on domain {args.domain}")
    print(f"  GUID: {dp.guid_prefix}")
    print(f"  Topic: {topic.name}")
    print("Waiting for data... Press Ctrl+C to stop.\n")

    # Send initial SPDP announcement
    dp.announce_spdp()

    try:
        while True:
            dp.spin_once(timeout=0.1)
    except KeyboardInterrupt:
        print("\nStopping...")
    finally:
        dp.stop()


if __name__ == "__main__":
    main()

File: rust/vibedds/examples/hello_sub.rs
/// HelloWorld subscriber example - receives HelloWorld messages
/// from hello_pub or rtiddsspy.
///
/// Run: cargo run --example hello_sub
/// Then in another terminal: cargo run --example hello_pub

use std::time::Duration;

use vibedds::participant::DomainParticipant;
use vibedds::qos::QosPolicy;
use vibedds::type_support::HelloWorldType;

fn main() {
    // Initialize logging
    env_logger::Builder::from_env(env_logger::Env::default().default_filter_or("info"))
        .format_timestamp_millis()
        .init();

    println!("VibeDDS HelloWorld Subscriber");
    println!("==============================");
    println!("Waiting for HelloWorld messages...");
    println!("Run 'cargo run --example hello_pub' to publish.");
    println!();

    // Create participant
    let mut participant = DomainParticipant::new(0, 1); // participant_id=1 to avoid port conflicts
    participant.set_spdp_announce_interval(Duration::from_secs(5));

    // Start the participant
    participant.start().expect("Failed to start participant");
    println!("Participant started: {:?}", participant.guid_prefix());

    // Send initial SPDP announcement
    participant.announce_spdp().expect("Failed to announce SPDP");

    // Create a HelloWorld DataReader
    let reader_qos = QosPolicy::reliable();
    let reader_arc = participant.create_reader(
        "HelloWorld",
        "HelloWorld",
        reader_qos,
    );
    println!("Created HelloWorld reader");

    // Set up data callback
    {
        let mut reader = reader_arc.lock().unwrap();
        reader.on_data(|payload| {
            match HelloWorldType::deserialize(payload) {
                Ok(message) => println!("Received: {}", message),
                Err(e) => log::error!("Failed to deserialize: {}", e),
            }
        });
    }

    // Set up discovery callback
    participant.on_participant_discovered(|p| {
        println!(
            "Discovered participant: {:?}, vendor={:?}",
            p.guid_prefix, p.vendor_id
        );
    });

    println!();
    println!("Waiting for data... Press Ctrl+C to stop.");
    println!();

    // Run the event loop
    loop {
        if let Err(e) = participant.spin_once(Duration::from_millis(100)) {
            log::error!("Error in spin: {}", e);
        }

        // Periodically announce SEDP
        if let Err(e) = participant.announce_sedp() {
            log::error!("Error in SEDP announce: {}", e);
        }

        // Also check the buffer for any samples not delivered via callback
        {
            let mut reader = reader_arc.lock().unwrap();
            while let Some(sample) = reader.take_one() {
                match HelloWorldType::deserialize(&sample.data) {
                    Ok(message) => {
                        println!(
                            "[buffer] Received: {} (from {:?}, sn={})",
                            message,
                            sample.writer_guid.prefix,
                            sample.sequence_number.value()
                        );
                    }
                    Err(e) => log::error!("Failed to deserialize: {}", e),
                }
            }
        }

        std::thread::sleep(Duration::from_millis(50));
    }
}
